File: rerun_py/rerun_sdk/rerun/components/view_coordinates_ext.py
from __future__ import annotations

from enum import IntEnum
from typing import TYPE_CHECKING, Any, Iterable, cast

import numpy as np
import numpy.typing as npt
import pyarrow as pa

if TYPE_CHECKING:
    from .._log import ComponentBatchLike
    from . import ViewCoordinates, ViewCoordinatesArrayLike


class ViewCoordinatesExt:
    class ViewDir(IntEnum):
        Up = 1
        Down = 2
        Right = 3
        Left = 4
        Forward = 5
        Back = 6

    @staticmethod
    def coordinates__field_converter_override(data: npt.ArrayLike) -> npt.NDArray[np.uint8]:
        coordinates = np.asarray(data, dtype=np.uint8)
        if coordinates.shape != (3,):
            raise ValueError(f"ViewCoordinates must be a 3-element array. Got: {coordinates.shape}")
        return coordinates

    @staticmethod
    def native_to_pa_array_override(data: ViewCoordinatesArrayLike, data_type: pa.DataType) -> pa.Array:
        from . import ViewCoordinates

        if isinstance(data, ViewCoordinates):
            # ViewCoordinates
            data = [data.coordinates]
        elif hasattr(data, "__len__") and len(data) > 0 and isinstance(data[0], ViewCoordinates):  # type: ignore[arg-type, index]
            # [ViewCoordinates]
            data = [d.coordinates for d in data]  # type: ignore[union-attr]
        else:
            try:
                # [x, y, z]
                data = [ViewCoordinates(data).coordinates]
            except ValueError:
                # [[x, y, z], ...]
                data = [ViewCoordinates(d).coordinates for d in data]  # type: ignore[union-attr]

        data = np.asarray(data, dtype=np.uint8)

        if len(data.shape) != 2 or data.shape[1] != 3:
            raise ValueError(f"ViewCoordinates must be a 3-element array. Got: {data.shape}")

        data = data.flatten()

        for value in data:
            # TODO(jleibs): Enforce this validation based on ViewDir
            if value not in range(1, 7):
                raise ValueError("ViewCoordinates must contain only values in the range [1,6].")

        return pa.FixedSizeListArray.from_arrays(data, type=data_type)

    # Implement the AsComponents protocol
    def as_component_batches(self) -> Iterable[ComponentBatchLike]:
        from ..archetypes import ViewCoordinates
        from ..components import ViewCoordinates as ViewCoordinatesComponent

        return ViewCoordinates(cast(ViewCoordinatesComponent, self)).as_component_batches()

    def num_instances(self) -> int:
        # Always a mono-component
        return 1

    # <BEGIN_GENERATED:declarations>
    # This section is generated by running `scripts/generate_view_coordinate_defs.py --python`
    # The following declarations are replaced in `deferred_patch_class`.
    ULF: ViewCoordinates = None  # type: ignore[assignment]
    UFL: ViewCoordinates = None  # type: ignore[assignment]
    LUF: ViewCoordinates = None  # type: ignore[assignment]
    LFU: ViewCoordinates = None  # type: ignore[assignment]
    FUL: ViewCoordinates = None  # type: ignore[assignment]
    FLU: ViewCoordinates = None  # type: ignore[assignment]
    ULB: ViewCoordinates = None  # type: ignore[assignment]
    UBL: ViewCoordinates = None  # type: ignore[assignment]
    LUB: ViewCoordinates = None  # type: ignore[assignment]
    LBU: ViewCoordinates = None  # type: ignore[assignment]
    BUL: ViewCoordinates = None  # type: ignore[assignment]
    BLU: ViewCoordinates = None  # type: ignore[assignment]
    URF: ViewCoordinates = None  # type: ignore[assignment]
    UFR: ViewCoordinates = None  # type: ignore[assignment]
    RUF: ViewCoordinates = None  # type: ignore[assignment]
    RFU: ViewCoordinates = None  # type: ignore[assignment]
    FUR: ViewCoordinates = None  # type: ignore[assignment]
    FRU: ViewCoordinates = None  # type: ignore[assignment]
    URB: ViewCoordinates = None  # type: ignore[assignment]
    UBR: ViewCoordinates = None  # type: ignore[assignment]
    RUB: ViewCoordinates = None  # type: ignore[assignment]
    RBU: ViewCoordinates = None  # type: ignore[assignment]
    BUR: ViewCoordinates = None  # type: ignore[assignment]
    BRU: ViewCoordinates = None  # type: ignore[assignment]
    DLF: ViewCoordinates = None  # type: ignore[assignment]
    DFL: ViewCoordinates = None  # type: ignore[assignment]
    LDF: ViewCoordinates = None  # type: ignore[assignment]
    LFD: ViewCoordinates = None  # type: ignore[assignment]
    FDL: ViewCoordinates = None  # type: ignore[assignment]
    FLD: ViewCoordinates = None  # type: ignore[assignment]
    DLB: ViewCoordinates = None  # type: ignore[assignment]
    DBL: ViewCoordinates = None  # type: ignore[assignment]
    LDB: ViewCoordinates = None  # type: ignore[assignment]
    LBD: ViewCoordinates = None  # type: ignore[assignment]
    BDL: ViewCoordinates = None  # type: ignore[assignment]
    BLD: ViewCoordinates = None  # type: ignore[assignment]
    DRF: ViewCoordinates = None  # type: ignore[assignment]
    DFR: ViewCoordinates = None  # type: ignore[assignment]
    RDF: ViewCoordinates = None  # type: ignore[assignment]
    RFD: ViewCoordinates = None  # type: ignore[assignment]
    FDR: ViewCoordinates = None  # type: ignore[assignment]
    FRD: ViewCoordinates = None  # type: ignore[assignment]
    DRB: ViewCoordinates = None  # type: ignore[assignment]
    DBR: ViewCoordinates = None  # type: ignore[assignment]
    RDB: ViewCoordinates = None  # type: ignore[assignment]
    RBD: ViewCoordinates = None  # type: ignore[assignment]
    BDR: ViewCoordinates = None  # type: ignore[assignment]
    BRD: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_X_UP: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_X_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_Y_UP: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_Y_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_Z_UP: ViewCoordinates = None  # type: ignore[assignment]
    RIGHT_HAND_Z_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_X_UP: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_X_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_Y_UP: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_Y_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_Z_UP: ViewCoordinates = None  # type: ignore[assignment]
    LEFT_HAND_Z_DOWN: ViewCoordinates = None  # type: ignore[assignment]
    # <END_GENERATED:declarations>

    @staticmethod
    def deferred_patch_class(cls: Any) -> None:
        # <BEGIN_GENERATED:definitions>
        # This section is generated by running `scripts/generate_view_coordinate_defs.py --python`
        cls.ULF = cls([cls.ViewDir.Up, cls.ViewDir.Left, cls.ViewDir.Forward])
        cls.UFL = cls([cls.ViewDir.Up, cls.ViewDir.Forward, cls.ViewDir.Left])
        cls.LUF = cls([cls.ViewDir.Left, cls.ViewDir.Up, cls.ViewDir.Forward])
        cls.LFU = cls([cls.ViewDir.Left, cls.ViewDir.Forward, cls.ViewDir.Up])
        cls.FUL = cls([cls.ViewDir.Forward, cls.ViewDir.Up, cls.ViewDir.Left])
        cls.FLU = cls([cls.ViewDir.Forward, cls.ViewDir.Left, cls.ViewDir.Up])
        cls.ULB = cls([cls.ViewDir.Up, cls.ViewDir.Left, cls.ViewDir.Back])
        cls.UBL = cls([cls.ViewDir.Up, cls.ViewDir.Back, cls.ViewDir.Left])
        cls.LUB = cls([cls.ViewDir.Left, cls.ViewDir.Up, cls.ViewDir.Back])
        cls.LBU = cls([cls.ViewDir.Left, cls.ViewDir.Back, cls.ViewDir.Up])
        cls.BUL = cls([cls.ViewDir.Back, cls.ViewDir.Up, cls.ViewDir.Left])
        cls.BLU = cls([cls.ViewDir.Back, cls.ViewDir.Left, cls.ViewDir.Up])
        cls.URF = cls([cls.ViewDir.Up, cls.ViewDir.Right, cls.ViewDir.Forward])
        cls.UFR = cls([cls.ViewDir.Up, cls.ViewDir.Forward, cls.ViewDir.Right])
        cls.RUF = cls([cls.ViewDir.Right, cls.ViewDir.Up, cls.ViewDir.Forward])
        cls.RFU = cls([cls.ViewDir.Right, cls.ViewDir.Forward, cls.ViewDir.Up])
        cls.FUR = cls([cls.ViewDir.Forward, cls.ViewDir.Up, cls.ViewDir.Right])
        cls.FRU = cls([cls.ViewDir.Forward, cls.ViewDir.Right, cls.ViewDir.Up])
        cls.URB = cls([cls.ViewDir.Up, cls.ViewDir.Right, cls.ViewDir.Back])
        cls.UBR = cls([cls.ViewDir.Up, cls.ViewDir.Back, cls.ViewDir.Right])
        cls.RUB = cls([cls.ViewDir.Right, cls.ViewDir.Up, cls.ViewDir.Back])
        cls.RBU = cls([cls.ViewDir.Right, cls.ViewDir.Back, cls.ViewDir.Up])
        cls.BUR = cls([cls.ViewDir.Back, cls.ViewDir.Up, cls.ViewDir.Right])
        cls.BRU = cls([cls.ViewDir.Back, cls.ViewDir.Right, cls.ViewDir.Up])
        cls.DLF = cls([cls.ViewDir.Down, cls.ViewDir.Left, cls.ViewDir.Forward])
        cls.DFL = cls([cls.ViewDir.Down, cls.ViewDir.Forward, cls.ViewDir.Left])
        cls.LDF = cls([cls.ViewDir.Left, cls.ViewDir.Down, cls.ViewDir.Forward])
        cls.LFD = cls([cls.ViewDir.Left, cls.ViewDir.Forward, cls.ViewDir.Down])
        cls.FDL = cls([cls.ViewDir.Forward, cls.ViewDir.Down, cls.ViewDir.Left])
        cls.FLD = cls([cls.ViewDir.Forward, cls.ViewDir.Left, cls.ViewDir.Down])
        cls.DLB = cls([cls.ViewDir.Down, cls.ViewDir.Left, cls.ViewDir.Back])
        cls.DBL = cls([cls.ViewDir.Down, cls.ViewDir.Back, cls.ViewDir.Left])
        cls.LDB = cls([cls.ViewDir.Left, cls.ViewDir.Down, cls.ViewDir.Back])
        cls.LBD = cls([cls.ViewDir.Left, cls.ViewDir.Back, cls.ViewDir.Down])
        cls.BDL = cls([cls.ViewDir.Back, cls.ViewDir.Down, cls.ViewDir.Left])
        cls.BLD = cls([cls.ViewDir.Back, cls.ViewDir.Left, cls.ViewDir.Down])
        cls.DRF = cls([cls.ViewDir.Down, cls.ViewDir.Right, cls.ViewDir.Forward])
        cls.DFR = cls([cls.ViewDir.Down, cls.ViewDir.Forward, cls.ViewDir.Right])
        cls.RDF = cls([cls.ViewDir.Right, cls.ViewDir.Down, cls.ViewDir.Forward])
        cls.RFD = cls([cls.ViewDir.Right, cls.ViewDir.Forward, cls.ViewDir.Down])
        cls.FDR = cls([cls.ViewDir.Forward, cls.ViewDir.Down, cls.ViewDir.Right])
        cls.FRD = cls([cls.ViewDir.Forward, cls.ViewDir.Right, cls.ViewDir.Down])
        cls.DRB = cls([cls.ViewDir.Down, cls.ViewDir.Right, cls.ViewDir.Back])
        cls.DBR = cls([cls.ViewDir.Down, cls.ViewDir.Back, cls.ViewDir.Right])
        cls.RDB = cls([cls.ViewDir.Right, cls.ViewDir.Down, cls.ViewDir.Back])
        cls.RBD = cls([cls.ViewDir.Right, cls.ViewDir.Back, cls.ViewDir.Down])
        cls.BDR = cls([cls.ViewDir.Back, cls.ViewDir.Down, cls.ViewDir.Right])
        cls.BRD = cls([cls.ViewDir.Back, cls.ViewDir.Right, cls.ViewDir.Down])
        cls.RIGHT_HAND_X_UP = cls([cls.ViewDir.Up, cls.ViewDir.Right, cls.ViewDir.Forward])
        cls.RIGHT_HAND_X_DOWN = cls([cls.ViewDir.Down, cls.ViewDir.Right, cls.ViewDir.Back])
        cls.RIGHT_HAND_Y_UP = cls([cls.ViewDir.Right, cls.ViewDir.Up, cls.ViewDir.Back])
        cls.RIGHT_HAND_Y_DOWN = cls([cls.ViewDir.Right, cls.ViewDir.Down, cls.ViewDir.Forward])
        cls.RIGHT_HAND_Z_UP = cls([cls.ViewDir.Right, cls.ViewDir.Forward, cls.ViewDir.Up])
        cls.RIGHT_HAND_Z_DOWN = cls([cls.ViewDir.Right, cls.ViewDir.Back, cls.ViewDir.Down])
        cls.LEFT_HAND_X_UP = cls([cls.ViewDir.Up, cls.ViewDir.Right, cls.ViewDir.Back])
        cls.LEFT_HAND_X_DOWN = cls([cls.ViewDir.Down, cls.ViewDir.Right, cls.ViewDir.Forward])
        cls.LEFT_HAND_Y_UP = cls([cls.ViewDir.Right, cls.ViewDir.Up, cls.ViewDir.Forward])
        cls.LEFT_HAND_Y_DOWN = cls([cls.ViewDir.Right, cls.ViewDir.Down, cls.ViewDir.Back])
        cls.LEFT_HAND_Z_UP = cls([cls.ViewDir.Right, cls.ViewDir.Back, cls.ViewDir.Up])
        cls.LEFT_HAND_Z_DOWN = cls([cls.ViewDir.Right, cls.ViewDir.Forward, cls.ViewDir.Down])
        # <END_GENERATED:definitions>

File: rerun_cpp/src/rerun/components/view_coordinates_ext.cpp
#include "view_coordinates.hpp"

// <CODEGEN_COPY_TO_HEADER>
#include "../rerun_sdk_export.hpp"

// </CODEGEN_COPY_TO_HEADER>

// Uncomment for better auto-complete while editing the extension.
// #define EDIT_EXTENSION

namespace rerun {
    namespace components {

#ifdef EDIT_EXTENSION
        struct ViewCoordinatesExt {
            uint8_t coordinates[3];
#define ViewCoordinates ViewCoordinatesExt

            // <CODEGEN_COPY_TO_HEADER>

            enum ViewDir : uint8_t {
                Up = 1,
                Down = 2,
                Right = 3,
                Left = 4,
                Forward = 5,
                Back = 6,
            };

            /// Construct Vec3D from x/y/z values.
            constexpr ViewCoordinates(uint8_t axis0, uint8_t axis1, uint8_t axis2)
                : coordinates{axis0, axis1, axis2} {}

            /// Construct Vec3D from x/y/z values.
            constexpr ViewCoordinates(ViewDir axis0, ViewDir axis1, ViewDir axis2)
                : coordinates{axis0, axis1, axis2} {}

            // <BEGIN_GENERATED:declarations>
            // This section is generated by running `scripts/generate_view_coordinate_defs.py --cpp`
            /// X=Up, Y=Left, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates ULF;

            /// X=Up, Y=Forward, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates UFL;

            /// X=Left, Y=Up, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LUF;

            /// X=Left, Y=Forward, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LFU;

            /// X=Forward, Y=Up, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FUL;

            /// X=Forward, Y=Left, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FLU;

            /// X=Up, Y=Left, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates ULB;

            /// X=Up, Y=Back, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates UBL;

            /// X=Left, Y=Up, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LUB;

            /// X=Left, Y=Back, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LBU;

            /// X=Back, Y=Up, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BUL;

            /// X=Back, Y=Left, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BLU;

            /// X=Up, Y=Right, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates URF;

            /// X=Up, Y=Forward, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates UFR;

            /// X=Right, Y=Up, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RUF;

            /// X=Right, Y=Forward, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RFU;

            /// X=Forward, Y=Up, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FUR;

            /// X=Forward, Y=Right, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FRU;

            /// X=Up, Y=Right, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates URB;

            /// X=Up, Y=Back, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates UBR;

            /// X=Right, Y=Up, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RUB;

            /// X=Right, Y=Back, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RBU;

            /// X=Back, Y=Up, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BUR;

            /// X=Back, Y=Right, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BRU;

            /// X=Down, Y=Left, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DLF;

            /// X=Down, Y=Forward, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DFL;

            /// X=Left, Y=Down, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LDF;

            /// X=Left, Y=Forward, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LFD;

            /// X=Forward, Y=Down, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FDL;

            /// X=Forward, Y=Left, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FLD;

            /// X=Down, Y=Left, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DLB;

            /// X=Down, Y=Back, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DBL;

            /// X=Left, Y=Down, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LDB;

            /// X=Left, Y=Back, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LBD;

            /// X=Back, Y=Down, Z=Left
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BDL;

            /// X=Back, Y=Left, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BLD;

            /// X=Down, Y=Right, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DRF;

            /// X=Down, Y=Forward, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DFR;

            /// X=Right, Y=Down, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RDF;

            /// X=Right, Y=Forward, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RFD;

            /// X=Forward, Y=Down, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FDR;

            /// X=Forward, Y=Right, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates FRD;

            /// X=Down, Y=Right, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DRB;

            /// X=Down, Y=Back, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates DBR;

            /// X=Right, Y=Down, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RDB;

            /// X=Right, Y=Back, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RBD;

            /// X=Back, Y=Down, Z=Right
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BDR;

            /// X=Back, Y=Right, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates BRD;

            /// X=Up, Y=Right, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_X_UP;

            /// X=Down, Y=Right, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_X_DOWN;

            /// X=Right, Y=Up, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_Y_UP;

            /// X=Right, Y=Down, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_Y_DOWN;

            /// X=Right, Y=Forward, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_Z_UP;

            /// X=Right, Y=Back, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates RIGHT_HAND_Z_DOWN;

            /// X=Up, Y=Right, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_X_UP;

            /// X=Down, Y=Right, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_X_DOWN;

            /// X=Right, Y=Up, Z=Forward
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_Y_UP;

            /// X=Right, Y=Down, Z=Back
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_Y_DOWN;

            /// X=Right, Y=Back, Z=Up
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_Z_UP;

            /// X=Right, Y=Forward, Z=Down
            RERUN_SDK_EXPORT static const rerun::components::ViewCoordinates LEFT_HAND_Z_DOWN;

            // <END_GENERATED:declarations>

            // </CODEGEN_COPY_TO_HEADER>
        };
#endif
        // <BEGIN_GENERATED:definitions>
        // This section is generated by running `scripts/generate_view_coordinate_defs.py --cpp`
        const ViewCoordinates ViewCoordinates::ULF = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::UFL = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::LUF = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::LFU = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::FUL = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::FLU = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::ULB = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::UBL = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::LUB = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::LBU = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::BUL = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::BLU = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::URF = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::UFR = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::RUF = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::RFU = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::FUR = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::FRU = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::URB = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::UBR = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::RUB = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::RBU = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::BUR = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::BRU = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::DLF = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::DFL = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::LDF = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::LFD = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::FDL = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::FLD = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::DLB = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::DBL = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::LDB = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::LBD = ViewCoordinates(
            rerun::components::ViewCoordinates::Left, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::BDL = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Left
        );
        const ViewCoordinates ViewCoordinates::BLD = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Left,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::DRF = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::DFR = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::RDF = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::RFD = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::FDR = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::FRD = ViewCoordinates(
            rerun::components::ViewCoordinates::Forward, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::DRB = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::DBR = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::RDB = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::RBD = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::BDR = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Right
        );
        const ViewCoordinates ViewCoordinates::BRD = ViewCoordinates(
            rerun::components::ViewCoordinates::Back, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_X_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_X_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_Y_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_Y_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_Z_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::RIGHT_HAND_Z_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Down
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_X_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Up, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_X_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Down, rerun::components::ViewCoordinates::Right,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_Y_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Up,
            rerun::components::ViewCoordinates::Forward
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_Y_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Down,
            rerun::components::ViewCoordinates::Back
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_Z_UP = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Back,
            rerun::components::ViewCoordinates::Up
        );
        const ViewCoordinates ViewCoordinates::LEFT_HAND_Z_DOWN = ViewCoordinates(
            rerun::components::ViewCoordinates::Right, rerun::components::ViewCoordinates::Forward,
            rerun::components::ViewCoordinates::Down
        );
        // <END_GENERATED:definitions>

    } // namespace components
} // namespace rerun
